tests/test-cli.cpp

```diff
@@ -0,0 +1,76 @@
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+
+#include <algorithm>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#include <unistd.h>
+
+static std::string read(const std::string & file) {
+    std::ostringstream actuals;
+    actuals << std::ifstream(file.c_str()).rdbuf();
+    return actuals.str();
+}
+
+static void assert_equals(const std::string & expected, const std::string & actual) {
+    if (expected != actual) {
+        std::cerr << "Expected: " << expected << std::endl;
+        std::cerr << "Actual: " << actual << std::endl;
+        std::cerr << std::flush;
+        throw std::runtime_error("Test failed");
+    }
+}
+
+static void assert_contains(const std::string & expected, const std::string & actual) {
+    if (actual.find(expected) == std::string::npos) {
+        std::cerr << "Expected to find: " << expected << std::endl;
+        std::cerr << "Actual: " << actual << std::endl;
+        std::cerr << std::flush;
+        throw std::runtime_error("Test failed");
+    }
+}
+
+struct Out {
+    std::string out;
+    std::string err;
+};
+
+static Out run(const std::string & cmd) {
+    auto full_cmd = cmd + " > out/out.txt 2> out/err.txt";
+    std::cerr << "Running: " << full_cmd << std::endl;
+    if (std::system(full_cmd.c_str()) != 0)
+        throw std::runtime_error("llama-cli binary failed to run.");
+    return {
+        /* .out = */ read("out/out.txt"),
+        /* .err = */ read("out/err.txt"),
+    };
+}
+
+int main(int argc, char ** argv) {
+    std::string cli_bin = argc == 2 ? argv[1] : "./llama-cli";
+
+    std::system("mkdir out/");
+
+    {
+        auto p = run(cli_bin + " --help");
+        if (!p.err.empty())
+            throw std::runtime_error("llama-cli --help should not have any stderr.");
+        assert_contains("example usage", p.out);
+    }
+
+    {
+        auto p = run(cli_bin + " -hfr ggml-org/models -hff tinyllamas/stories260K.gguf --prompt hello --seed 42 -ngl 0 -n 10");
+        assert_equals(" hello Joe and Joe we", p.out);
+        assert_contains("system_info:", p.err);
```

Comment:
Default sampling settings can change in the future, using greedy sampling here may be more reliable.